pgjdbc/src/test/java/org/postgresql/test/jdbc2/optional/BaseDataSourceFailoverUrlsTest.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2004, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.test.jdbc2.optional;
+
+import static org.junit.Assert.assertEquals;
+
+import org.postgresql.ds.common.BaseDataSource;
+
+import org.junit.Test;
+
+import java.io.IOException;
+import javax.naming.NamingException;
+
+/**
+* tests that failover urls survive the parse/rebuild roundtrip with and without specific ports
+*/
+public class BaseDataSourceFailoverUrlsTest {
```

Comment:
I don't like the equals on regex splits. It's really checking for a prefix so make that explicit via a helper and invoke the helper (ex: `assertStartsWith(...)`).

The helper `doTestFailoverUrl(...)` should create separate data sources for each style of assignment it is testing rather than clobbering the same reference. So something like this:


```java
public void testFullDefault() {
    roundTripFromUrl("jdbc:postgresql://server/database", "jdbc:postgresql://server:5432/database");
}

// ...

@Test
public void testNullPorts() {
    BaseDataSource bds = newDS();
    bds.setDatabaseName("database");
    bds.setPortNumbers(null);
    assertStartsWith(bds.getUrl(), "jdbc:postgresql://localhost/database");
    assertEquals(0, bds.getPortNumber());
    assertEquals(0, bds.getPortNumbers()[0]);
}

// ...

private void roundTripFromUrl(String url, String expectedUrlPrefix) {
    BaseDataSource bdsFromUrl = newDS();
    bdsFromUrl.setUrl(url);
    assertStartsWith(bdsFromUrl.getUrl(), expectedUrlPrefix);

    BaseDataSource bdsFromRef = newDS();
    bdsFromRef.setFromReference(bdsFromUrl.getReference());
    assertStartsWith(bdsFromRef.getUrl(), expectedUrlPrefix);

    BaseDataSource bdsFromInit = newDS();
    bdsFromInit.initializeFrom(bdsFromUrl);
    assertStartsWith(bdsFromInit.getUrl(), expectedUrlPrefix);
}

private static void assertStartsWith(String value, String prefix) {
    // ...
}
```

Reply:
I don't understand all you're saying.
I introduced the assert method you mentioned.
(but it's not "startsWith" but "equalsUntilQuestionmark", and "expected" should come first in assert, so I didn't completely copy over your suggestion).
I'm also fine if you refactor and extend the tests after this PR, if you have something broader in mind.